### DISCUSSION TITLE
Signature coordinates error on different operating systems

### DISCUSSION THREAD

**Discussion by haumenphai** [MatthiasValvekens/pyHanko#475] [Q&A]
I tested pyhanko on Window and Ubuntu,
I have a question why the coordinates of signature field are different.

for example with coordinates:
```python
signature_coords = {'x1': 158.30325, 'y1': 591.17825, 'x2': 278.30325000000005, 'y2': 666.1782499999999}
```

on Ubuntu  fill in the order x1,y1,x2,y2 and the result is as expected.

```python
                pdf_signer = signers.PdfSigner(
                    signature_meta=signers.PdfSignatureMetadata(
                        field_name=data['signature_field_name']
                    ),
                    new_field_spec=SigFieldSpec(
                        sig_field_name=data['signature_field_name'],
                        on_page=data['on_page'] - 1,
                        box=(
                            int(signature_coords['x1']),
                            int(signature_coords['y1']),
                            int(signature_coords['x2']),
                            int(signature_coords['y2'])
                        )
                    ),
         
                    signer=window_signer
                )
```

But on Windows, I have to fill in differently to get the correct box like on Ubuntu.
```python
box=(
    int(signature_coords['x1']),
    int(signature_coords['y2']),
    int(signature_coords['x2']),
    int(signature_coords['y1'])
)
```

I want my code to run well on different operating systems, this makes me wonder.

===
version: 0.25.1

**Comment by niranjanpathak-np1** [comment]
**Adjusting Coordinates Based on Platform**
You can programmatically detect the platform using Python's os module and adjust the coordinates accordingly. For example:

**Detect the Operating System:**

```
python

import os
platform = os.name  # 'posix' for Ubuntu, 'nt' for Windows
```
Coordinate Transformation (if needed): If the origin or orientation differs, you might need to flip the Y-axis or adjust the coordinates accordingly. For example, if the Y-axis is inverted on Windows, you could modify the y1 and y2 values.

Here is a generalized approach:

```
python

if platform == 'nt':  # Windows
    # Assuming the Y-axis is inverted on Windows
    pdf_height = 842  # Example for an A4 page height in points (adjust accordingly)
    signature_coords['y1'] = pdf_height - signature_coords['y1']
    signature_coords['y2'] = pdf_height - signature_coords['y2']
```
Use Adjusted Coordinates in the Box: Now you can fill the box with the adjusted coordinates:
```

python

box=(
    int(signature_coords['x1']),
    int(signature_coords['y1']),
    int(signature_coords['x2']),
    int(signature_coords['y2'])
)
```
**Example Code with Platform Check:**
```
python

import os

# Signature coordinates example
signature_coords = {
    'x1': 158.3025,
    'y1': 591.17825,
    'x2': 278.30235,
    'y2': 666.17825
}

```
# Detect platform
`platform = os.name  # 'posix' for Ubuntu, 'nt' for Windows`

# Adjust coordinates based on platform (example adjustment for Windows)
```
if platform == 'nt':  # Windows
    pdf_height = 842  # Example A4 page height in points, adjust as per your PDF
    signature_coords['y1'] = pdf_height - signature_coords['y1']
    signature_coords['y2'] = pdf_height - signature_coords['y2']
```

# Fill the box with adjusted coordinates
```
box = (
    int(signature_coords['x1']),
    int(signature_coords['y1']),
    int(signature_coords['x2']),
    int(signature_coords['y2'])
)
```

# Now use this box for creating the signature field as in your original code
```
pdf_signer = signers.PdfSigner(
    signature_meta=signers.PdfSignatureMetadata(
        field_name_data=['signature_field_name']
    ),
    new_field_spec=SigFieldSpec(
        sig_field_name_data='signature_field_name',
        on_page_data=['on_page'] - 1,
        box=box
    ),
    signer=window_signer
)
```
**Explanation:**
Platform Detection: This allows the script to know whether it's running on Windows or Ubuntu.
Y-axis Adjustment: If the Y-axis is inverted on Windows (common in some systems), the y1 and y2 coordinates are recalculated by subtracting them from the page height (pdf_height), essentially flipping the Y-axis.
Box Filling: After the coordinates are adjusted, the signature field is placed using the modified coordinates to ensure consistent behavior across platforms.

**Comment by haumenphai** [comment]
Yes, I sloved that problem, but i'm not sure if it always happens with windows os.

**Comment by MatthiasValvekens** [comment]
@niranjanpathak-np1 While your attempt at a solution is appreciated, please don't just blindly post LLM-generated answers to questions. _Especially_ when the LLM misses the point by a mile...

Anyway, this is not likely to be a pyHanko issue, but rather a "I have two PDF viewers that produce different results when fed the same file" issue. The difference in OSes is also not really relevant. PyHanko is completely platform-agnostic and performs the exact same PDF manipulations regardless of the platform on which it is executed.

Please include the names & versions of the viewers you're using, and also your input document, since this can only be pyHanko's fault if it somehow turned a valid PDF into an invalid one during the signing process.

I'll also be converting this into a discussion.
